src/Models/BaseElement.php

```diff
@@ -528,6 +528,49 @@ public function getContentForSearchIndex(): string
         return $content;
     }
 
+    /**
+     * Provides content for CMS search
+     */
+    public function getContentForCmsSearch(): string
+    {
+        $contents = [];
+        foreach ($this->config()->get('db') as $fieldName => $fieldType) {
+            // https://docs.silverstripe.org/en/developer_guides/model/data_types_and_casting/
+            // probably won't actually get things like 'CanViewType' in db config
+            if ($fieldName === 'LastEdited'
+                || $fieldName === 'Created'
+                || $fieldName === 'CanViewType'
+                || $fieldName === 'CanEditType'
+                || $fieldName === 'Version'
+                || $fieldName === 'ShowInMenus'
+                || $fieldName === 'ShowInSearch'
+                || $fieldName === 'Sort'
+                || $fieldName === 'HasBrokenFile'
+                || $fieldName === 'HasBrokenLink'
+                || $fieldName === 'ReportClass'
+                || substr($fieldType, -2) === 'ID'
+                || substr($fieldType, -3) === 'Key'
+                || substr($fieldName, -9) === 'ClassName'
+                || substr($fieldType, -4) === 'Hash'
+            ) {
+                continue;
+            }
+            // TODO: if $fieldType is HTMLText then run shortcode parser on it?
+            // (possibly not? unlikely to really matter from a cms search perspective?)
+            $contents[] = $this->$fieldName;
+        }
+        // Use |#| to delimite different fields rather than space so that you don't
+        // accidentally join results of two columns that are next to each other in a table
+        $content = implode('|#|', $contents);
+
+        // Strips tags and be sure there's a space between words.
+        $content = trim(strip_tags(str_replace('<', ' <', $content)));
+
+        // Allow projects to update content of third-party elements.
+        $this->extend('updateContentForCmsSearch', $content);
```

Comment:
That could well make sense, though if we use  a string it allows projects to do useful things like this:

```php
    protected function updateContentForCmsSearch(string &$content)
    {
        $content .= 'ヽ(⌐■_■)ノ♪♬';
    }
```

Reply:
I think we need the extension points because we need a way how to remove things as well, not just how to add them.

Reply:
Is removing the content when it's already in this string okay though? If I was developing with this I'd much rather have control over which _fields_ are included, knowing that fields A, B, and C are relevant but fields D and E aren't. I guess it doesn't hurt to have both ways of controlling it, but this _does_ make it so we can't later optimise this option further by having it do the filter in the DB directly as discussed in https://github.com/silverstripe/silverstripe-elemental/pull/1069#discussion_r1244608200

Reply:
Yeah, my gut feel is to have the extension and allow modifications on the string. Could we want e.g. different logic based on subsites? Or other stuff where a single config might not be flexible enough?

Reply:
> I'd much rather have control over which fields are included

As I've mentioned previously, there already is `quasi-config` in that the "config of fields that are included" is currently "whatever fields are included in the template".  Obviously we can't use that here.

I'm pretty against the idea of opt-in config for the fields you want to include. I want to actually get adoption on this and I view "having to define every single field on every single element" as a pretty big annoyance.  Also it'll degrade over time because any time a new developer adds a new content block (very standard for junior devs) then it's super easy to forget / not know that you need to add the new field to the CMS SiteTree search thing as well.

I think if we're going to having configurable fields, best we can do here is create a configurable list of fields to NOT include

Reply:
I've added config array `BaseElement.exclude_fields_from_cms_search` to allow for opt-ing out fields

Reply:
I've decided I'm okay with searching across all db fields with configurable exceptions - and it doesn't hurt to have this extension point, so that's fine too.

But I _would_ ideally also like a way to add extra fields which are part of relations, since it is _very_ common to have a block which gets its content almost exclusively from relations (e.g. [carousels](https://github.com/dnadesign/silverstripe-elemental-carousel), [nested blocks](https://github.com/dnadesign/silverstripe-elemental-list), [decision trees](https://github.com/dnadesign/silverstripe-elemental-decisiontree), [accordions](https://github.com/dynamic/silverstripe-elemental-accordion), etc etc).

@michalkleiner I'd like your input on that side as well since you've expressed interest in the configuration aspects of this PR - would you find it useful to have the ability to add relation fields into the list of searchable fields?

Reply:
Yes, with the "index the rendered template" approach that content is included, so the change here shouldn't make the content scope smaller, or should provide an alternative way how to achieve the same. Alt tags for linked images, nested carousel content etc.

The extension point is good-enough affordance for that, so if we go with "all content db fields less configured exclusion fields plus content provided/removed via the extension point", then that'd work for us.

If we could support some form of wildcard mechanism for the exclusions that'd be dope, e.g. `Video*` would exclude all fields starting with 'Video' (`VideoLink`, `VideoTitle`...), '*' would exclude all and let the extension handle what's needed specifically on project level.

Reply:
`Video*` seems almost trivial to do without a wildcard - it's easy to exclude fields on a per element class level with the config in this PR:

```yml
App\MyElement:
  fields_excluded_from_cms_search:
    - MyFieldToExclude
    - AnotherFieldToExclude
```

I think for simplicity sake it better to NOT include a wildcard mechanism, I don't think it's consistent with any other config in Silverstripe

Re adding fields from related classes, that's easy enough to do using the new extension point:

// extension is applied to any Element class in the project
```php
class MyElementExtension extends DataExtension
{
    protected function updateContentForCmsSearch(string &content)
    {
        $content .= $this->owner->MyRelation()->MyField;
    }
}
```

Reply:
> Re adding fields from related classes, that's easy enough to do using the new extension point

It is a bit awkward to have to update a string when you just want to add a field - and in your example the delimiter is missed out which means your newly added field will be appended directly to the end of the previous field which I think is going to be a common mistake. That said, this PR has had about enough time in the oven and this would be an easy enhancement to make in the future if people want it, so I'm okay leaving the extensibility as is for now.